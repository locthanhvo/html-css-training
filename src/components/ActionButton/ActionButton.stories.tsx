import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import ActionButton from '.';

import { AddIcon, ForwardIcon } from '../Icons';

export default {
  title: 'Components/ActionButton',
  component: ActionButton,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof ActionButton>;

export const SignInButton: Story = {
  args: {
    size: 'xl',
    title: 'Sign In',
    bgColor: 'blue.200',
    rightIcon: <ForwardIcon />,
  },
};

export const SignUpButton: Story = {
  args: {
    size: 'xl',
    title: 'Sign Up',
    bgColor: 'blue.200',
  },
};

export const AddUserButton: Story = {
  args: {
    size: 'lg',
    title: 'Add User',
    bgColor: 'blue.200',
    rightIcon: <AddIcon />,
  },
};

export const EditButton: Story = {
  args: {
    size: 'sm',
    title: 'Edit',
    bgColor: 'white',
    color: 'blue.100',
    borderColor: 'blue.100',
  },
};

export const DeleteButton: Story = {
  args: {
    size: 'sm',
    title: 'Delete',
    bgColor: 'red.900',
  },
};

export const CancelButton: Story = {
  args: {
    size: 'lg',
    title: 'Cancel',
    color: 'gray.500',
    borderColor: 'gray.300',
  },
};
