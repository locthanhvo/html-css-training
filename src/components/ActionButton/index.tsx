import { JSXElementConstructor, ReactElement, memo } from 'react';
import { Button } from '@chakra-ui/react';

interface ActionButtonProps {
  type?: 'button' | 'submit' | 'reset';
  size?: 'sm' | 'md' | 'lg' | 'xl';
  isLoading?: boolean;
  isDisabled?: boolean;
  title: string;
  bgColor?: string;
  color?: string;
  borderColor?: string;
  onClick?: () => void;
  rightIcon?:
    | ReactElement<any, string | JSXElementConstructor<any>>
    | undefined;
}

const ActionButton = ({
  title,
  type = 'button',
  size,
  isLoading,
  isDisabled,
  bgColor,
  color,
  borderColor,
  rightIcon,
  onClick,
}: ActionButtonProps) => {
  return (
    <Button
      type={type}
      name={title}
      size={size}
      bgColor={bgColor}
      color={color}
      border="1px solid"
      borderColor={borderColor}
      onClick={onClick}
      rightIcon={rightIcon}
      isLoading={isLoading}
      isDisabled={isDisabled}
    >
      {title}
    </Button>
  );
};

export default memo(ActionButton);
