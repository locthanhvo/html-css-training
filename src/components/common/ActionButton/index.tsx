import { memo } from 'react';
import { Button, ButtonProps } from '@chakra-ui/react';

interface ActionButtonProps extends ButtonProps {
  type?: 'button' | 'submit' | 'reset';
  size?: 'sm' | 'md' | 'lg' | 'xl';
  isLoading?: boolean;
  isDisabled?: boolean;
  title?: string;
  bgColor?: string;
  color?: string;
  borderColor?: string;
  onClick?: () => void;
}

const ActionButton = ({
  title = 'Save',
  type = 'button',
  size = 'sm',
  isLoading = false,
  isDisabled = false,
  bgColor,
  color,
  borderColor,
  onClick,
  ...props
}: ActionButtonProps) => {
  return (
    <Button
      type={type}
      name={title}
      size={size}
      bgColor={bgColor}
      color={color}
      border="1px solid"
      borderColor={borderColor}
      onClick={onClick}
      isLoading={isLoading}
      isDisabled={isDisabled}
      {...props}
    >
      {title}
    </Button>
  );
};

export default memo(ActionButton);
