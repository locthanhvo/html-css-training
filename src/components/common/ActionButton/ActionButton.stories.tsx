import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import ActionButton from '.';

export default {
  title: 'Components/ActionButton',
  component: ActionButton,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof ActionButton>;

export const Default: Story = {
  args: {
    size: 'sm',
    title: 'Save',
    bgColor: 'royalBlue',
  },
};

export const EditButton: Story = {
  args: {
    size: 'md',
    title: 'Edit',
    bgColor: 'royalBlue',
  },
};

export const CancelButton: Story = {
  args: {
    size: 'md',
    title: 'Cancel',
    border: '1px solid',
    color: 'coolGray',
  },
};

export const DeleteButton: Story = {
  args: {
    size: 'md',
    title: 'Delete',
    border: '1px solid',
    color: 'lightRed',
  },
};
