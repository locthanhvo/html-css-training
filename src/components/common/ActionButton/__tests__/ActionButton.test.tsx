import { RenderResult, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import ActionButton from '..';

const mockOnClick = jest.fn();

describe('Action Button Component', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <BrowserRouter>
        <ActionButton bgColor="royalBlue" onClick={mockOnClick} />
      </BrowserRouter>,
    );
  });

  it('should match Button Save component', () => {
    const { container } = renderResult;

    expect(container).toMatchSnapshot();
  });

  it('should call onClick', () => {
    const { getByText } = renderResult;

    getByText('Save').click();
    expect(mockOnClick).toHaveBeenCalled();
  });
});
