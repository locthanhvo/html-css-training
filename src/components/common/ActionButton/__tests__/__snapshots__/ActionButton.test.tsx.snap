// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Button Component should match Button Save component 1`] = `
<div>
  <button
    class="chakra-button css-1exh3ze"
    name="Save"
    type="button"
  >
    Save
  </button>
</div>
`;
